xline/src/storage/leasestore/lease_heap.rs

```diff
@@ -0,0 +1,128 @@
+use std::{
+    cmp::Ordering,
+    collections::{BinaryHeap, HashMap},
+};
+
+use tokio::time::Instant;
+
+/// Lease id and expiration time
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub(crate) struct LeaseWithTime {
+    /// Lease id
+    lease_id: i64,
+    /// Expiration time
+    expiry: Instant,
+}
+
+impl LeaseWithTime {
+    /// New `LeaseWithTime`
+    pub(crate) fn new(lease_id: i64, expiry: Instant) -> Self {
+        Self { lease_id, expiry }
+    }
+
+    /// Check if lease is expired
+    pub(crate) fn expired(&self) -> bool {
+        self.expiry <= Instant::now()
+    }
+
+    /// Lease id
+    pub(crate) fn id(&self) -> i64 {
+        self.lease_id
+    }
+}
+
+impl PartialOrd for LeaseWithTime {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for LeaseWithTime {
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.expiry.cmp(&self.expiry)
+    }
+}
+
+/// Updatable binary heap of lease
+#[derive(Debug)]
+pub(crate) struct LeaseHeap {
+    /// Backup of lease heap, used to update lease heap
+    map: HashMap<i64, LeaseWithTime>,
+    /// Lease heap
+    heap: BinaryHeap<LeaseWithTime>,
+}
+
+impl LeaseHeap {
+    /// New `LeaseHeap`
+    pub(crate) fn new() -> Self {
+        Self {
+            map: HashMap::new(),
+            heap: BinaryHeap::new(),
+        }
+    }
+
+    /// Insert or update lease
+    pub(crate) fn insert_or_update(&mut self, le: LeaseWithTime) {
+        if self.map.insert(le.lease_id, le).is_some() {
+            self.heap = self.map.values().copied().collect();
```

Comment:
`pop_first` was not stable until version 1.66. In addition, LeaseWithTime is sorted by expiry and uses lease id as the unique id. BTreeSet can only replace elements with the same expiry (because it is sorted by expiry inside the set), even if it can replace elements with the same lease id, because expire updates, the order is also shuffled, so re-sort is always necessary

Reply:
Take a look at this crate https://docs.rs/priority-queue/latest/priority_queue/. It supports changing priority and may help us get rid of `LeaseWithTime`.

Reply:
This is a better option, I have refactored it